packages/@lwc/synthetic-shadow/src/polyfills/mutation-observer/polyfill.ts

```diff
@@ -115,68 +114,69 @@ function filterMutationRecords(
                     ) {
-                        ArrayPush.call(filteredSet, record);
-                    } else if (shadowRoot) {
-                        const shadowRootObservers = getNodeObservers(shadowRoot);
+                        ArrayPush.call(result, record);
+                    } else {
+                        // else, must be observing the shadowRoot
 
-                        if (
-                            shadowRootObservers &&
-                            (shadowRootObservers[0] === observer ||
-                                ArrayIndexOf.call(shadowRootObservers, observer) !== -1)
-                        ) {
-                            ArrayPush.call(filteredSet, retargetMutationRecord(record));
-                        }
+                        ArrayPush.call(result, retargetMutationRecord(record));
                     }
                 }
             } else {
-                // Mutation happened under a root node(shadow root or document) and the decision is straighforward
-                // Ascend the tree starting from target and check if observer is qualified
-                if (isQualifiedObserver(observer, target)) {
-                    ArrayPush.call(filteredSet, record);
+                const { removedNodes } = record;
+                // In the case of removed nodes, climbing the tree is not an option as the nodes are disconnected
+                // We can only check if either the host or shadow root was observed and qualify the record
+                const shadowRoot = (target as Element).shadowRoot;
+                const sampleNode = removedNodes[0];
+                if (
+                    getNodeNearestOwnerKey(target) === getNodeNearestOwnerKey(sampleNode) && // trickery: sampleNode is slot content
+                    isQualifiedObserver(observer, target) // use target as a close enough reference to climb up
+                ) {
+                    ArrayPush.call(result, record);
+                } else if (shadowRoot) {
+                    const shadowRootObservers = getNodeObservers(shadowRoot);
+
+                    if (
+                        shadowRootObservers &&
+                        (shadowRootObservers[0] === observer ||
+                            ArrayIndexOf.call(shadowRootObservers, observer) !== -1)
+                    ) {
+                        ArrayPush.call(result, retargetMutationRecord(record));
+                    }
                 }
             }
-            return filteredSet;
-        },
-        []
-    ) as MutationRecord[];
+        } else {
+            // Mutation happened under a root node(shadow root or document) and the decision is straighforward
+            // Ascend the tree starting from target and check if observer is qualified
+            if (isQualifiedObserver(observer, target)) {
+                ArrayPush.call(result, record);
+            }
+        }
+    }
+    return result;
```

Comment:
This diff is hard to read, but there are 2 small changes here:

1. Convert `ArrayReduce` into a regular for-loop (this is the bulk of the improvement).
2. Lazily access properties from the `record` rather than all at once (this is maybe a 1% perf improvement – apparently the browser does some work when you call these getters).